File: sdks/node/src/query_builder.ts
/**
 * Fluent query builder for listing rows in an Applad table.
 *
 * Created via {@link Databases.from} — do not instantiate directly.
 *
 * @example
 * ```ts
 * const result = await client.databases
 *   .from('myDb', 'posts')
 *   .select('id, title, author(name)')
 *   .equal('published', true)
 *   .orderDesc('created_at')
 *   .limit(25)
 *   .get();
 *
 * console.log(result.total);  // total matching rows
 * console.log(result.rows);   // the page of rows
 * ```
 */

export interface QueryResult {
  total: number;
  rows: Record<string, unknown>[];
}

type RowFetcher = (params: Record<string, unknown>) => Promise<QueryResult>;

export class QueryBuilder {
  private readonly _fetch: RowFetcher;
  private _queries: string[] = [];
  private _select?: string;
  private _orderAttr?: string;
  private _orderType?: string;
  private _limit?: number;
  private _offset?: number;
  private _cursorAfter?: string;

  /** @internal */
  constructor(fetch: RowFetcher) {
    this._fetch = fetch;
  }

  // ── Column selection ──────────────────────────────────────────────────────

  /** Specify which columns to return, e.g. `'id, title, author(name)'`. */
  select(columns: string): this {
    this._select = columns;
    return this;
  }

  // ── Filters ───────────────────────────────────────────────────────────────

  /** Matches rows where `field` equals `value`. */
  equal(field: string, value: unknown): this {
    return this._scalar('equal', field, value);
  }

  /** Matches rows where `field` does not equal `value`. */
  notEqual(field: string, value: unknown): this {
    return this._scalar('notEqual', field, value);
  }

  /** Matches rows where `field` is less than `value`. */
  lessThan(field: string, value: unknown): this {
    return this._scalar('lessThan', field, value);
  }

  /** Matches rows where `field` is less than or equal to `value`. */
  lessThanOrEqual(field: string, value: unknown): this {
    return this._scalar('lessThanEqual', field, value);
  }

  /** Matches rows where `field` is greater than `value`. */
  greaterThan(field: string, value: unknown): this {
    return this._scalar('greaterThan', field, value);
  }

  /** Matches rows where `field` is greater than or equal to `value`. */
  greaterThanOrEqual(field: string, value: unknown): this {
    return this._scalar('greaterThanEqual', field, value);
  }

  /** Matches rows where `field` contains `value` (case-insensitive). */
  contains(field: string, value: string): this {
    return this._scalar('contains', field, value);
  }

  /** Matches rows where `field` starts with `value`. */
  startsWith(field: string, value: string): this {
    return this._scalar('startsWith', field, value);
  }

  /** Matches rows where `field` ends with `value`. */
  endsWith(field: string, value: string): this {
    return this._scalar('endsWith', field, value);
  }

  /** Full-text search on `field` for `value`. */
  search(field: string, value: string): this {
    return this._scalar('search', field, value);
  }

  /** Matches rows where `field` is NULL. */
  isNull(field: string): this {
    this._queries.push(`isNull("${field}")`);
    return this;
  }

  /** Matches rows where `field` is NOT NULL. */
  isNotNull(field: string): this {
    this._queries.push(`isNotNull("${field}")`);
    return this;
  }

  /** Matches rows where `field` is between `min` and `max` (inclusive). */
  between(field: string, min: unknown, max: unknown): this {
    this._queries.push(`between("${field}","${min}","${max}")`);
    return this;
  }

  // ── Ordering ──────────────────────────────────────────────────────────────

  /** Order results by `field` ascending. */
  orderAsc(field: string): this {
    this._orderAttr = field;
    this._orderType = 'ASC';
    return this;
  }

  /** Order results by `field` descending. */
  orderDesc(field: string): this {
    this._orderAttr = field;
    this._orderType = 'DESC';
    return this;
  }

  // ── Pagination ────────────────────────────────────────────────────────────

  /** Maximum number of rows to return. */
  limit(n: number): this {
    this._limit = n;
    return this;
  }

  /** Number of rows to skip (offset-based pagination). */
  offset(n: number): this {
    this._offset = n;
    return this;
  }

  /**
   * Cursor-based pagination — pass the last seen row ID to fetch the next page.
   *
   * @example
   * ```ts
   * const page1 = await builder.limit(25).get();
   * const page2 = await builder.limit(25).cursorAfter(page1.rows.at(-1)!['$id'] as string).get();
   * ```
   */
  cursorAfter(rowId: string): this {
    this._cursorAfter = rowId;
    return this;
  }

  // ── Execution ─────────────────────────────────────────────────────────────

  /** Execute the query and return a {@link QueryResult}. */
  async get(): Promise<QueryResult> {
    return this._fetch(this._buildParams());
  }

  // ── Internal ──────────────────────────────────────────────────────────────

  private _scalar(method: string, field: string, value: unknown): this {
    this._queries.push(`${method}("${field}","${value}")`);
    return this;
  }

  private _buildParams(): Record<string, unknown> {
    const params: Record<string, unknown> = {};
    if (this._select !== undefined) params['select'] = this._select;
    if (this._orderAttr !== undefined) params['orderAttr'] = this._orderAttr;
    if (this._orderType !== undefined) params['orderType'] = this._orderType;
    if (this._limit !== undefined) params['limit'] = this._limit;
    if (this._offset !== undefined) params['offset'] = this._offset;
    if (this._cursorAfter !== undefined) params['cursorAfter'] = this._cursorAfter;
    if (this._queries.length > 0) params['queries[]'] = this._queries;
    return params;
  }
}

File: sdks/dart/lib/src/query_builder.dart
/// A fluent query builder for listing rows in an Applad table.
///
/// Created via [Databases.from] — do not instantiate directly.
///
/// ```dart
/// final result = await client.databases
///     .from('myDb', 'posts')
///     .select('id, title, author(name)')
///     .equal('published', true)
///     .orderDesc('created_at')
///     .limit(25)
///     .get();
///
/// print(result.total);   // total matching rows
/// print(result.rows);    // List<Map<String, dynamic>>
/// ```
library;

/// The fetch function type used internally by [QueryBuilder].
typedef RowFetcher = Future<Map<String, dynamic>> Function(
    Map<String, dynamic> params);

class QueryBuilder {
  final RowFetcher _fetch;
  final List<String> _queries = [];

  String? _select;
  String? _orderAttr;
  String? _orderType;
  int? _limit;
  int? _offset;
  String? _cursorAfter;

  /// Creates a [QueryBuilder]. Prefer [Databases.from] over calling this directly.
  QueryBuilder(RowFetcher fetch) : _fetch = fetch;

  // ── Column selection ────────────────────────────────────────────────────────

  /// Specify which columns to return, e.g. `'id, title, author(name)'`.
  /// Use `author(name)` syntax to select a related column via a relationship.
  QueryBuilder select(String columns) {
    _select = columns;
    return this;
  }

  // ── Filters ─────────────────────────────────────────────────────────────────

  /// Matches rows where [field] equals [value].
  QueryBuilder equal(String field, dynamic value) =>
      _scalar('equal', field, value);

  /// Matches rows where [field] does not equal [value].
  QueryBuilder notEqual(String field, dynamic value) =>
      _scalar('notEqual', field, value);

  /// Matches rows where [field] is less than [value].
  QueryBuilder lessThan(String field, dynamic value) =>
      _scalar('lessThan', field, value);

  /// Matches rows where [field] is less than or equal to [value].
  QueryBuilder lessThanOrEqual(String field, dynamic value) =>
      _scalar('lessThanEqual', field, value);

  /// Matches rows where [field] is greater than [value].
  QueryBuilder greaterThan(String field, dynamic value) =>
      _scalar('greaterThan', field, value);

  /// Matches rows where [field] is greater than or equal to [value].
  QueryBuilder greaterThanOrEqual(String field, dynamic value) =>
      _scalar('greaterThanEqual', field, value);

  /// Matches rows where [field] contains [value] (case-insensitive LIKE).
  QueryBuilder contains(String field, String value) =>
      _scalar('contains', field, value);

  /// Matches rows where [field] starts with [value].
  QueryBuilder startsWith(String field, String value) =>
      _scalar('startsWith', field, value);

  /// Matches rows where [field] ends with [value].
  QueryBuilder endsWith(String field, String value) =>
      _scalar('endsWith', field, value);

  /// Full-text search on [field] for [value].
  QueryBuilder search(String field, String value) =>
      _scalar('search', field, value);

  /// Matches rows where [field] is NULL.
  QueryBuilder isNull(String field) {
    _queries.add('isNull("$field")');
    return this;
  }

  /// Matches rows where [field] is NOT NULL.
  QueryBuilder isNotNull(String field) {
    _queries.add('isNotNull("$field")');
    return this;
  }

  /// Matches rows where [field] is between [min] and [max] (inclusive).
  QueryBuilder between(String field, dynamic min, dynamic max) {
    _queries.add('between("$field","$min","$max")');
    return this;
  }

  // ── Ordering ─────────────────────────────────────────────────────────────────

  /// Order results by [field] ascending.
  QueryBuilder orderAsc(String field) {
    _orderAttr = field;
    _orderType = 'ASC';
    return this;
  }

  /// Order results by [field] descending.
  QueryBuilder orderDesc(String field) {
    _orderAttr = field;
    _orderType = 'DESC';
    return this;
  }

  // ── Pagination ───────────────────────────────────────────────────────────────

  /// Maximum number of rows to return. Defaults to the server default (25).
  QueryBuilder limit(int n) {
    _limit = n;
    return this;
  }

  /// Number of rows to skip (offset-based pagination).
  /// Prefer [cursorAfter] for large datasets.
  QueryBuilder offset(int n) {
    _offset = n;
    return this;
  }

  /// Cursor-based pagination — pass the last seen row ID to fetch the next page.
  ///
  /// ```dart
  /// // First page
  /// final page1 = await builder.limit(25).get();
  ///
  /// // Next page
  /// final page2 = await builder
  ///     .limit(25)
  ///     .cursorAfter(page1.rows.last['$id'])
  ///     .get();
  /// ```
  QueryBuilder cursorAfter(String rowId) {
    _cursorAfter = rowId;
    return this;
  }

  // ── Execution ────────────────────────────────────────────────────────────────

  /// Execute the query and return a [QueryResult].
  Future<QueryResult> get() async {
    final data = await _fetch(_buildParams());
    return QueryResult._fromJson(data);
  }

  // ── Internal ─────────────────────────────────────────────────────────────────

  QueryBuilder _scalar(String method, String field, dynamic value) {
    _queries.add('$method("$field","$value")');
    return this;
  }

  Map<String, dynamic> _buildParams() {
    final params = <String, dynamic>{};
    if (_select != null) params['select'] = _select;
    if (_orderAttr != null) params['orderAttr'] = _orderAttr;
    if (_orderType != null) params['orderType'] = _orderType;
    if (_limit != null) params['limit'] = _limit;
    if (_offset != null) params['offset'] = _offset;
    if (_cursorAfter != null) params['cursorAfter'] = _cursorAfter;
    if (_queries.isNotEmpty) params['queries[]'] = _queries;
    return params;
  }
}

/// The result of a [QueryBuilder.get] call.
class QueryResult {
  /// Total number of rows matching the query (before limit/offset).
  final int total;

  /// The rows returned for this page.
  final List<Map<String, dynamic>> rows;

  const QueryResult({required this.total, required this.rows});

  factory QueryResult._fromJson(Map<String, dynamic> json) {
    final rawRows = json['rows'];
    final rows = rawRows is List
        ? rawRows
            .map((r) => Map<String, dynamic>.from(r as Map))
            .toList()
        : <Map<String, dynamic>>[];
    return QueryResult(
      total: (json['total'] as num?)?.toInt() ?? rows.length,
      rows: rows,
    );
  }

  @override
  String toString() => 'QueryResult(total: $total, rows: ${rows.length})';
}
